util/provenance/buildinfo.go

```diff
@@ -0,0 +1,164 @@
+package provenance
+
+import (
+	"encoding/hex"
+	"strings"
+
+	distreference "github.com/docker/distribution/reference"
+	slsa "github.com/in-toto/in-toto-golang/in_toto/slsa_provenance/v0.2"
+	binfotypes "github.com/moby/buildkit/util/buildinfo/types"
+	digest "github.com/opencontainers/go-digest"
+	ocispecs "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+var BuildKitBuildType = "https://mobyproject.org/buildkit@v1"
+
+type ProvenancePredicate struct {
+	slsa.ProvenancePredicate
+	Metadata *ProvenanceMetadata                `json:"metadata,omitempty"`
+	Source   *Source                            `json:"buildSource,omitempty"`
+	Layers   map[string][][]ocispecs.Descriptor `json:"buildLayers,omitempty"`
+}
+
+type ProvenanceMetadata struct {
+	slsa.ProvenanceMetadata
+	VCS map[string]string `json:"vcs,omitempty"`
```

Comment:
So this creates a new `vcs` in the provenance `metadata`?

I wonder if we should follow https://slsa.dev/provenance/v0.2:

> Producers MAY add extension fields using field names that are URIs.

Or alternatively, we could add it to the materials? If I read the spec correctly, it looks like adding a new key `vcs` wouldn't be allowed, and we'd possibly be breaking compatibility?

Reply:
Indeed probably safer to do the URI. There will be many fields like this that are non-spec (5-10).

One of the bad ones is that SLSA does not make a distinction between "materials complete" and "hermetic".

Reply:
These are the current non-spec values:

```
vcs metadata

buildSource
buildLayers
```

```
Completeness.Hermetic

list of local sources
list of secrets
list of ssh
list of frontend inputs

future:
exporter options (these can probably be in parameters)
deps array of subpredictes for chained builds
```

Build steps are under `buildconfig` that is `interface{}` type. I think we should try to group them under a same field as much as possible. Parameters is also an `interface{}` type, but I think better to keep it only for string map so it is predictable.

Reply:
https://github.com/tonistiigi/buildkit/commit/6c6e6768c73beba35bc431be48d2475580fc272b#diff-ef30221db67d0b6659402bd38285c0932c7c5beef2b99e4c34b92c546a64b9acR17 wdyt?

Reply:
SGTM :tada:

I wonder if maybe VCS content might be better suited to be in the environment rather than the metadata, given that it is an input to the build?

Reply:
If you actually build from Git then it is under invocation. But if build was actually from local source then I think it is better to mark it as such under input info and only put the vcs as "additional metadata".

Reply:
Oh ok, as long as we're noting that it's built from local source somewhere in the environment/input/etc, that's good :+1: